Sources/Prometheus/MetricTypes/Histogram.swift

```diff
@@ -123,20 +123,27 @@ public class Histogram<NumType: DoubleRepresentable, Labels: HistogramLabels>: M
     /// - Parameters:
     ///     - value: Value to observe
     ///     - labels: Labels to attach to the observed value
-    public func observe(_ value: NumType, _ labels: Labels? = nil) {
+    ///     - done: Completion handler
+    ///     - observedValue: Value written to the histogram
+    public func observe(_ value: NumType, _ labels: Labels? = nil, _ done: @escaping (_ observedValue: NumType) -> Void = { _ in }) {
         prometheusQueue.async(flags: .barrier) {
             if let labels = labels, type(of: labels) != type(of: EmptySummaryLabels()) {
                 let his = self.prometheus.getOrCreateHistogram(with: labels, for: self)
-                his.observe(value)
+                his.observe(value, nil, done)
             }
             self.total.inc(value)
             
             for (i, bound) in self.upperBounds.enumerated() {
                 if bound >= value.doubleValue {
                     self.buckets[i].inc()
-                    return
+                    break
                 }
             }
+            
+            // Only run the callback once; not on both the label and without labels
+            if labels == nil {
```

Comment:
This will not work. If you pass in labels, `done` will be called before the part after the `if let labels` is executed.

This should be changed to something like this:
```swift
if let labels = labels ... {
    let his = ...
    his.observe(value)
}
self.total.inc(value)

for (i, bound) in ... {
    ...
}
done()
```
making sure done is only called (if passed in by a user) once the instance the user called on is fully done.

Reply:
Yup, 100% the right move; thanks!